lib/email/sender.rb

```diff
@@ -232,6 +232,18 @@ def send
       email_log
     end
 
+    def find_user
+      return @user if @user
+      if @email_type.to_s == "confirm_new_email"
+        user = EmailChangeRequest.find_by(
+          new_email: to_address,
+          change_state: EmailChangeRequest.states[:authorizing_new]
+        )&.user
```

Comment:
What happens if there are multiple `EmailChangeRequest` rows? I can imagine it happening like

- admin1 changes email to test@example.com
- admin1 changes email to test2@example.com
- admin2 changes email to test@example.com

Now there will be two EmailChangeRequest rows for test@example.com. One for `admin1`, and one for `admin2`. I believe `find_by` will always return the record with the lowest `id`, which may cause problems?

Reply:
Wow :facepalm: did not even think of that. Thanks for pointing this out, will try think of something else. I am not 100% sure I need to do this, can maybe just add the email type to the list of allowed bypass types.

Reply:
@davidtaylorhq I think this is fine now, I just skip looking up the EmailChangeRequest altogether, it is not necessary because `@user` is always present for UserEmails which the confirm email change one definitely falls under.

Reply:
Let's take a step back here. The text for the setting says:

> Prevent Discourse from sending any kind of emails. Select 'yes' to disable emails for all users. Select 'non-staff' to disable emails for non-staff users only.

The operative word here is `for`. By the time we get to the `Email::Sender`, we don't know who the instigator of this action is. We're assuming that the person that we are sending the email to is the same as the person that is being served, but they aren't necessarily the same, for example with invites.

Reply:
We might assume that emails are either send by staff or they are automatic, and if they are automatic, `@user` is set. In which case, we should never lookup a user by the to address.

Reply:
@danielwaterworth I am not sure what you would like me to do here? I don't see why we wouldn't look up the user by the to address to see if we are sending to a staff user IF the `@user` is not present? I would like to merge this particular PR because this process is currently broken.

Reply:
When is that actually necessary?

Reply:
@danielwaterworth there are a few places that call `Sender.new` and do not pass in a user:

```
app/jobs/regular/invite_email.rb Email::Sender.new(message, :invite).send
app/jobs/scheduled/version_check.rb Email::Sender.new(message, :new_version).send
app/jobs/regular/group_smtp_email.rb Email::Sender.new(message, :group_smtp).send                                                                                                                                               
app/jobs/regular/download_backup_email.rb Email::Sender.new(message, :download_backup_message).send                                                                                                                             
app/jobs/regular/admin_confirmation_email.rb Email::Sender.new(message, :admin_confirmation_message).send                                                                                                                       
app/jobs/regular/invite_password_instructions_email.rb Email::Sender.new(message, :invite_password_instructions).send
```

I am not sure if any of these warrant checking if the outgoing to address is an admin user? If not we could probably get rid of the finding User by `to_address`.

Reply:
Let's go through them in order:

 * For invites, I thought it was only staff that could send invites, but it's TL2 and above. Checking the to_address does not make sense here. I think the proper behaviour is only allowing staff to send invites when disable emails is set to non-staff,
 * Version check emails are sent to the `contact_email` site setting. I wouldn't expect these emails to stop even if the email was not associated with a staff account,
 * Group SMTP emails are tricky, I'm not sure what should happen,
 * Backups can only be triggered in the admin interface,
 * Admin confirmation emails are only sent to admins,
 * Invite password instructions emails are also tricky, it's up for interpretation,

Reply:
Hmm well I think I can just bypass the check for all of these using the BYPASS const except for invite password instructions, and make the change to the invite rules for staff sending. I think I’d rather make this second change in a new PR though because it’s more substantial. I’ll ping you once I’ve made the other changes.

Reply:
cc @danielwaterworth